pkg/kubevirt/client.go

```diff
@@ -285,44 +288,24 @@ func (c *client) getSnapshotClassNameFromVolumeClaimName(ctx context.Context, na
 		klog.V(2).Infof("Error getting storage class name for claim %s in namespace %s: %v", claimName, namespace, err)
 		return "", fmt.Errorf("unable to determine snapshot class name for infra source volume")
 	}
-	if storageClassName == "" {
-		return "", fmt.Errorf("unable to determine storage class name for snapshot creation")
-	}
-	allowed, err := c.isStorageClassAllowed(ctx, storageClassName)
-	if err != nil {
-		return "", err
+	if storageClassName == "" && !c.storageClassEnforcement.AllowDefault {
+		return "", fmt.Errorf("unable to determine snapshot class name for snapshot creation, and default not allowed")
+	} else if storageClassName != "" && !(util.Contains(c.storageClassEnforcement.AllowList, storageClassName) || c.storageClassEnforcement.AllowAll) {
+		return "", fmt.Errorf("unable to determine snapshot class name for snapshot creation, no valid snapshot classes found")
 	}
-	if !allowed {
-		return "", fmt.Errorf("storage class %s is not allowed for snapshot creation", storageClassName)
-	}
-	snapshotClass, err := c.getSnapshotClassFromStorageClass(ctx, storageClassName, snapshotClassName)
-	if err != nil {
-		return "", err
+	snapshotClassNames := c.storageClassEnforcement.GetInfraSnapshotClassesFromStorageClassName(storageClassName)
+	if util.Contains(snapshotClassNames, snapshotClassName) {
+		return snapshotClassName, nil
 	}
-	return snapshotClass.Name, nil
-}
-
-func (c *client) isStorageClassAllowed(ctx context.Context, storageClassName string) (bool, error) {
-	if !c.storageClassEnforcement.AllowAll && !util.Contains(c.storageClassEnforcement.AllowList, storageClassName) {
-		if c.storageClassEnforcement.AllowDefault {
-			// Check if storage class is default and default is allowed.
-			storageClass, err := c.kubernetesClient.StorageV1().StorageClasses().Get(ctx, storageClassName, metav1.GetOptions{})
-			if err != nil {
-				if errors.IsNotFound(err) {
-					return false, nil
-				} else {
-					return false, err
-				}
-			}
-			ann := storageClass.GetAnnotations()
-			if v, ok := ann["storageclass.kubernetes.io/is-default-class"]; !ok || v != "true" {
-				return false, nil
-			}
+	if !(c.storageClassEnforcement.AllowAll || c.storageClassEnforcement.AllowDefault) {
+		tenantSnapshotClasses := c.storageClassEnforcement.GetTenantSnapshotClassesFromInfraStorageClassName(storageClassName)
+		if len(tenantSnapshotClasses) > 0 {
+			return "", fmt.Errorf("unable to determine snapshot class name for snapshot creation, valid snapshot classes are %v", tenantSnapshotClasses)
 		} else {
-			return false, nil
+			return "", fmt.Errorf("unable to determine snapshot class name for snapshot creation, no valid snapshot classes found")
```

Comment:
It's possible users will find these error messages a bit confusing. If they explicitly set a volume snapshot class name, but get an error that says `unable to determine snapshot class name for snapshot creation` that wouldn't make a lot of sense.

in the case of incompatibility, i'd expect an error message that says somethign like `volume snapshot class X is not compatible with PVC with storageclass Y, valid snapshot classes for this pvc are %v`

Reply:
Okay let me reword that message a bit.